res/controllers/Roland_DJ-202-scripts.js

```diff
@@ -0,0 +1,1612 @@
+    const isLeftDeck = this.number <= 8;
+    const channel = isLeftDeck ? 0x94 : 0x95;
+    this.midi = [channel, 0x20 + this.number - (isLeftDeck ? 0 : 8)];
+    components.SamplerButton.prototype.send.call(this, value);
+    this.midi = [channel + 2, 0x20 + this.number - (isLeftDeck ? 0 : 8)];
+    components.SamplerButton.prototype.send.call(this, value);
+};
+
+DJ202.SamplerButton.prototype.unshift = function() {
+    this.input = function(_channel, _control, value, _status, _group) {
+        const isLeftDeck = this.number <= 8;
+        const padMode = isLeftDeck ? DJ202.leftDeck.padSection.mode :
+            DJ202.rightDeck.padSection.mode;
+        if (padMode === DJ202.PadMode.SLICER) {
+            DJ202.slicer(value, this.number, false);
+            return;
+        } else if (padMode === DJ202.PadMode.SLICERLOOP) {
+            DJ202.slicer(value, this.number, true);
+            return;
+        }
+        components.SamplerButton.prototype.input.apply(this, arguments);
+    };
+};
+
+/////
+
+DJ202.Sampler = function() {
+    components.ComponentContainer.call(this);
+    this.syncDeck = -1;
+    this.button = [];
+
+    for (let i = 1; i <= 16; i++) {
+        this.button[i] = new DJ202.SamplerButton({
+            sendShifted: true,
+            shiftControl: true,
+            shiftOffset: 8,
+            number: i
+        });
+    }
+
+    this.level = new components.Pot({
+        inValueScale: function(value) {
+            // FIXME: The sampler gain knob has a dead zone and appears to
+            // scale non-linearly.
+            return components.Pot.prototype.inValueScale.call(this, value) * 4;
+        },
+        input: function(_channel, _control, value, _status, _group) {
+            if (!DJ202.bindSamplerControls) {
+                return;
+            }
+            for (let i = 1; i <= 16; i++) {
+                const group = `[Sampler${  i  }]`;
+                engine.setValue(group, "pregain", this.inValueScale(value));
+            }
+        }
+    });
+
+    this.pfl = new components.Button({
+        sampler: this,
+        midi: [0x9f, 0x1d],
+        connect: function() {
+            if (!DJ202.bindSamplerControls) {
+                return;
+            }
+            components.Button.prototype.connect.call(this);
+            // Ensure a consistent state between mixxx and device.
+            for (let i = 1; i <= 16; i++) {
+                const group = `[Sampler${  i  }]`;
+                engine.setValue(group, "pfl", false);
+            }
+            this.send(0);
+        },
+        input: function(_channel, _control, value, _status, _group) {
+            if (!value || !DJ202.bindSamplerControls) {
+                return;
+            }
+            for (let i = 1; i <= 16; i++) {
+                const group = `[Sampler${  i  }]`;
+                script.toggleControl(group, "pfl");
+            }
+        }
+    });
+
+    // TR-S
+    ///////////////////////
```

Comment:
Please make this comment more clear - it looks incomplete

Reply:
Removed comment, meaningless anyway